mlflow/store/azure_blob_artifact_repo.py

```diff
@@ -94,12 +94,12 @@ def list_artifacts(self, path=None):
             results = self.client.list_blobs(container, prefix=prefix, delimiter='/', marker=marker)
             for r in results:
                 if isinstance(r, BlobPrefix):   # This is a prefix for items in a subdirectory
-                    subdir = r.name[len(artifact_path)+1:]
+                    subdir = os.path.relpath(path=r.name, start=artifact_path)
```

Comment:
is the stored `r.name` always a full path starting after `blob.core.windows.net/`? If not `os.path.relpath` gives wrong answers. 

```
>>> os.path.relpath("a/b/c/", "p/q/r")
'../../../a/b/c'
```

even this one..

```
>>> os.path.relpath("a/b/c/", "b/c/p/q/r")
'../../../../../a/b/c'
```

Let's assert that it does in this code. Also add a unit test for this case. So if this is broken in future we know...

Reply:
`r.name` is always a path containing the prefix specified by the `prefix` parameter of `self.client.list_blobs()`. In this case, we pass a prefix value of `prefix 
 = posixpath.join(artifact_path, <subdirectory_path>). 

We then take the relative path with `start=artifact_path`. Because `prefix.startswith(artifact_path)`, this should work in all documented cases. For reference, see the documentation for the `prefix` parameter of the [azure storage `list_blobs` Python API method](https://azure-storage.readthedocs.io/ref/azure.storage.blob.baseblobservice.html#azure.storage.blob.baseblobservice.BaseBlobService.list_blobs).

I've added a check to verify that the listed blob's have names beginning with the specified prefix. Let's discuss the structure of the additional unit test!

Reply:
Added a test case to ensure that the new exception is raised when listed blob's do not contain the artifact path prefix!